### PR TITLE
Add utility options to Dask DataFrame type loader/materializer

### PR DESCRIPTION
**Pull request by kinghuang**
This PR contains the other half of #2809, and is meant to be merged after #2811.

Add configuration options to the Dask DataFrame loader and materializer configs to perform some common operations on DataFrames. The operations are:

* drop
* sample
* reset_index
* set_index
* repartition
* normalize_column_names

The first five correspond directly to methods of the same name on DataFrame objects. Multiple utilities can be set simultaneously. They are evaluated in the order of the list, above.

While solids can directly perform these operations on DataFrames themselves, incorporating them into the DataFrame type loader and materializer helps keep solids focussed on their main business logic and be adaptable to common data conditions. Here's a few example scenarios.

1. Systems like Amazon Athena do not support upper case column names. Materialize datasets with broadly compatible column names by setting `normalize_column_names: true` to lowercase names and convert CamelCase to snake_case.

```yaml
solids:
  df_solid:
    outputs:
      - output_df:
          to:
            parquet:
              path: s3://bucket/dataset
          normalize_column_names: true
```

2. The number of partitions in a DataFrame influences how much Dask can scale an operation across workers. A CSV file that is read as 1 partition can be easily repartitioned to 12 by setting `repartition: {"n_partitions": 12}`.

```yaml
solids:
  df_solid:
    inputs:
      input_df:
        from:
          csv:
            path: /project/data.csv
        repartition:
          npartitions: 12
```

3. Running a complex pipeline can take a long time. Sample 0.1% of the input as a quick preview by setting `sample: {"frac": 0.011}` before running the pipeline on a full dataset.

```yaml
solids:
  df_solid:
    inputs:
      input_df:
        from:
          csv:
            path: /project/data.csv
        sample:
          frac: 0.001
```

These utilities are meant to cover common needs during loading and materialization, only. They are not a replacement for more complex operations in solids.

### CODE REVIEW

**Comment by kinghuang** [comment]
Putting this back to draft to prevent merging. PR #2811 should go in first.

File: python_modules/libraries/dagster-dask/dagster_dask/data_frame.py [hunk 1]
```diff
@@ -395,7 +419,18 @@ def dataframe_loader(_context, config):
     read_args = [read_options.pop("path")] if read_meta.get("is_path_based", False) else []
     read_kwargs = read_options
 
+    # Convert frozenlist to list due to frozenlist being unpickle-able.
+    # https://github.com/dagster-io/dagster/issues/2719
+    read_args = [list(value) if isinstance(value, frozenlist) else value for value in read_args]
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-dask/dagster_dask/data_frame.py:424]: this is resolved

File: python_modules/libraries/dagster-dask/dagster_dask/data_frame.py [hunk 2]
```diff
@@ -350,12 +364,22 @@ def _dataframe_loader_config():
...
+                util_name: util_spec["options"]
+                for util_name, util_spec in DataFrameUtilities.items()
+            },
             # https://github.com/dagster-io/dagster/issues/2872
             **{
-                field_name: Field(field_config, is_required=False,)
+                field_name: Field(
+                    field_config,
+                    is_required=False,
+                )
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-dask/dagster_dask/data_frame.py:382]: should we add some sort of test or invariant to ensure we don't accidentally overwrite keys here or [1]

File: python_modules/libraries/dagster-dask/dagster_dask/data_frame.py [hunk 3]
```diff
@@ -413,12 +448,22 @@ def _dataframe_materializer_config():
...
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-dask/dagster_dask/data_frame.py:466]: [1]

File: python_modules/libraries/dagster-dask/dagster_dask/data_frame.py (hunk 1 above)
Comment (kinghuang) [reply python_modules/libraries/dagster-dask/dagster_dask/data_frame.py:424]: Will remove.

File: python_modules/libraries/dagster-dask/dagster_dask/data_frame.py (hunk 2 above)
Comment (kinghuang) [reply python_modules/libraries/dagster-dask/dagster_dask/data_frame.py:382]: Fair point, especially with the new keys I'm adding. I'll split out the addition of the deprecated keys and check that they don't overlap.

Comment (kinghuang) [reply python_modules/libraries/dagster-dask/dagster_dask/data_frame.py]: Added two unit tests.

**Comment by mgasner** [comment]
this looks great, going to merge it -- 🙏  thank you!

**Pull request merged**